sn_service_management/src/node.rs

```diff
@@ -0,0 +1,283 @@
+use async_trait::async_trait;
+use libp2p::{multiaddr::Protocol, Multiaddr, PeerId};
+use serde::{de::Error as DeError, Deserialize, Deserializer, Serialize, Serializer};
+use service_manager::{ServiceInstallCtx, ServiceLabel};
+use sn_protocol::get_port_from_multiaddr;
+use std::{ffi::OsString, net::SocketAddr, path::PathBuf, str::FromStr};
+
+pub struct NodeService {
+    pub service_data: NodeServiceData,
+    pub rpc_actions: Box<dyn RpcActions + Send>,
+}
+
+impl NodeService {
+    pub fn new(
+        service_data: NodeServiceData,
+        rpc_actions: Box<dyn RpcActions + Send>,
+    ) -> NodeService {
+        NodeService {
+            rpc_actions,
+            service_data,
+        }
+    }
+}
+
+#[async_trait]
+impl ServiceStateActions for NodeService {
+    fn bin_path(&self) -> PathBuf {
+        self.service_data.safenode_path.clone()
+    }
+
+    fn build_upgrade_install_context(&self, options: UpgradeOptions) -> Result<ServiceInstallCtx> {
+        let label: ServiceLabel = self.service_data.service_name.parse()?;
+        let mut args = vec![
+            OsString::from("--rpc"),
+            OsString::from(self.service_data.rpc_socket_addr.to_string()),
+            OsString::from("--root-dir"),
+            OsString::from(
+                self.service_data
+                    .data_dir_path
+                    .to_string_lossy()
+                    .to_string(),
+            ),
+            OsString::from("--log-output-dest"),
+            OsString::from(self.service_data.log_dir_path.to_string_lossy().to_string()),
+        ];
+
+        if self.service_data.genesis {
+            args.push(OsString::from("--first"));
+        }
+        if self.service_data.local {
+            args.push(OsString::from("--local"));
+        }
+        if let Some(node_port) = self.service_data.get_safenode_port() {
+            args.push(OsString::from("--port"));
+            args.push(OsString::from(node_port.to_string()));
+        }
+
+        if !options.bootstrap_peers.is_empty() {
+            let peers_str = options
+                .bootstrap_peers
+                .iter()
+                .map(|peer| peer.to_string())
+                .collect::<Vec<_>>()
+                .join(",");
+            args.push(OsString::from("--peer"));
+            args.push(OsString::from(peers_str));
+        }
+
+        let mut service_ctx = ServiceInstallCtx {
+            label: label.clone(),
+            program: self.service_data.safenode_path.to_path_buf(),
+            args,
+            contents: None,
+            username: Some(self.service_data.user.to_string()),
+            working_directory: None,
+            environment: options.env_variables,
+        };
+        // Temporary fix to enable the restart cmd to properly restart a running service.
+        // 'ServiceInstallCtx::content' will override the other passed in fields.
+        #[cfg(target_os = "linux")]
```

Comment:
I think we can now reomve this override. The Daemon is responsible for restarting a service.